File: data/java/src/com/db/data/format/AviStreamHeader.java
/*
 * Copyright (c) 2007 Digital Bazaar, Inc.  All rights reserved.
 */
package com.db.data.format;

import java.io.IOException;
import java.io.OutputStream;

/**
 * An AVI Stream Header ('strh').
 * 
 * AVI Format is as follows:
 * 
 * AVI Form Header ('RIFF' size 'AVI ' data)
 *    Header List ('LIST' size 'hdrl' data)
 *       AVI Header ('avih' size data)
 *       Video Stream Header List ('LIST' size 'strl' data)
 *          Video Stream Header ('strh' size data)
 *          Video Stream Format ('strf' size data)
 *          Video Stream Data ('strd' size data) - IGNORED, for DRIVERS
 *       Audio Stream Header List ('LIST' size 'strl' data)
 *          Audio Stream Header ('strh' size data)
 *          Audio Stream Format ('strf' size data)
 *          Audio Stream Data ('strd' size data) - IGNORED, for DRIVERS
 *    Info List ('LIST' size 'INFO' data)
 *       Index Entry ({'ISBJ','IART','ICMT',...} size data)
 *    Movie List ('LIST' size 'movi' data)
 *       Movie Entry ({'00db','00dc','01wb'} size data)
 *    Index Chunk ('idx1' size data)
 *       Index Entry ({'00db','00dc','01wb',...})
 * 
 * ---------------------------------------------------------------
 * In a Stream Header 'strh' (FOURCC means a four-character code):
 * ---------------------------------------------------------------
 * FOURCC type - 'vids' for a video stream, 'auds' for an audio stream
 * FOURCC handler - the installable compressor or decompressor for the data
 * DWORD flags
 * DWORD reserved
 * DWORD initial frames
 * DWORD scale
 * DWORD rate
 * DWORD start
 * DWORD length
 * DWORD suggested buffer size
 * DWORD quality
 * DWORD sample size
 * 
 * The flags for the Stream Header:
 * 
 * AVISF_DISABLED - whether or not the data should only be rendered when
 * explicitly enabled by the user
 * 
 * AVISF_VIDEO_PALCHANGES - whether or not palette changes are embedded
 * in the file (chunks tagged like '00pc')
 * 
 * @author Dave Longley
 */
public class AviStreamHeader
{
   /**
    * The AVI stream header RIFF header.
    */
   protected RiffChunkHeader mRiffHeader;
   
   /**
    * The data for this chunk, not including its header.
    */
   protected byte[] mData;
   
   /**
    * Constructs a new AviStreamHeader.
    */
   public AviStreamHeader()
   {
      // create RIFF header
      mRiffHeader = new RiffChunkHeader("strh");
   }
   
   /**
    * Writes this AviStreamHeader, including the RIFF header, to an
    * OutputStream.
    * 
    * @param os the OutputStream to write to.
    * 
    * @exception IOException thrown if an IO error occurs.
    */
   public void writeTo(OutputStream os) throws IOException
   {
      // write RIFF header
      mRiffHeader.writeTo(os);
      
      // write stream data
      os.write(mData);
   }   
   
   /**
    * Returns whether or not this AviStreamHeader is valid.
    * 
    * @return true if valid, false if not.
    */
   public boolean isValid()
   {
      return mRiffHeader.isValid();
   }
   
   /**
    * Gets the size of this AviStreamHeader, not including its chunk header.
    * 
    * @return the size of this AviStreamHeader.
    */
   public long getSize()
   {
      return mRiffHeader.getChunkSize();
   }
}

File: data/cpp/db/data/avi/AviStreamHeader.cpp
/*
 * Copyright (c) 2007-2009 Digital Bazaar, Inc. All rights reserved.
 */
#include "db/data/avi/AviStreamHeader.h"

#include <cstring>

using namespace db::data;
using namespace db::data::avi;
using namespace db::data::riff;
using namespace db::io;

AviStreamHeader::AviStreamHeader() :
   mRiffHeader(CHUNK_ID)
{
}

AviStreamHeader::~AviStreamHeader()
{
}

bool AviStreamHeader::writeTo(OutputStream& os)
{
   bool rval;
   
   // write RIFF header
   rval = mRiffHeader.writeTo(os);
   
   if(rval)
   {
      // write data
      rval = os.write(mData, HEADER_SIZE);
   }
   
   return rval;
}

bool AviStreamHeader::convertFromBytes(const char* b, int length)
{
   bool rval = false;
   
   // convert the RIFF header
   if(mRiffHeader.convertFromBytes(b, length) && isValid())
   {
      // make sure length has enough data for the chunk
      if(length >= getSize())
      {
         memcpy(mData, b + RiffChunkHeader::HEADER_SIZE, getChunkSize());
         
         // converted successfully
         rval = true;
      }
   }
   
   return rval;
}

bool AviStreamHeader::isValid()
{
   return mRiffHeader.isValid() && mRiffHeader.getIdentifier() == CHUNK_ID;
}

int AviStreamHeader::getChunkSize()
{
   // AVI stream header size is 56 bytes
   return (int)mRiffHeader.getChunkSize();
}

int AviStreamHeader::getSize()
{
   return getChunkSize() + RiffChunkHeader::HEADER_SIZE;
}
